File: android/EGETrainerApp/core/src/main/kotlin/ru/egetrainer/core/models/AuthModels.kt
package ru.egetrainer.core.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Сессия Supabase Auth (ответ /auth/v1/token).
 * GoTrue может отдать expires_at ИЛИ только expires_in — после декода
 * обязательно вызывать [normalized] (зеркало init(from:) в AuthModels.swift).
 */
@Serializable
data class AuthSession(
    @SerialName("access_token") val accessToken: String,
    @SerialName("refresh_token") val refreshToken: String,
    @SerialName("expires_at") val expiresAt: Double? = null,
    @SerialName("expires_in") val expiresIn: Double? = null,
    val user: AuthUser,
) {
    /** expires_at гарантированно заполнен (из expires_in при отсутствии). */
    fun normalized(nowEpochSeconds: Double = System.currentTimeMillis() / 1000.0): AuthSession =
        if (expiresAt != null) this
        else copy(expiresAt = nowEpochSeconds + (expiresIn ?: 3600.0))

    val isExpiringSoon: Boolean
        get() = (expiresAt ?: 0.0) - System.currentTimeMillis() / 1000.0 < 60
}

@Serializable
data class AuthUser(
    val id: String,
    val email: String? = null,
)

/** Строка таблицы profiles (читается под RLS своим токеном). */
@Serializable
data class Profile(
    val id: String,
    val email: String? = null,
    val role: String? = null, // "student" | "teacher"
    @SerialName("first_name") val firstName: String? = null,
    @SerialName("last_name") val lastName: String? = null,
    @SerialName("student_grade") val studentGrade: Int? = null,
    @SerialName("teacher_type") val teacherType: String? = null,
    @SerialName("profile_completed") val profileCompleted: Boolean? = null,
) {
    val isTeacher: Boolean get() = role == "teacher"

    /** Профиль требует completion-шага (как редирект на google_complete.html). */
    val needsCompletion: Boolean
        get() {
            val r = (role ?: "").trim()
            if (r != "student" && r != "teacher") return true
            return profileCompleted == false
        }

    val displayName: String
        get() {
            val name = listOfNotNull(firstName, lastName).joinToString(" ").trim()
            if (name.isNotEmpty()) return name
            email?.substringBefore('@')?.takeIf { it.isNotEmpty() }?.let { return it }
            return "Пользователь"
        }

    /** Имя для start_homework_attempt — как inferNameFromUser в вебе. */
    val hwStudentName: String
        get() {
            firstName?.trim()?.takeIf { it.isNotEmpty() }?.let { return it }
            email?.substringBefore('@')?.takeIf { it.isNotEmpty() }?.let { return it }
            return "Ученик"
        }
}

File: ios/EGETrainerApp/EGETrainerApp/Models/AuthModels.swift
import Foundation

/// Сессия Supabase Auth (ответ /auth/v1/token).
struct AuthSession: Codable, Equatable {
    var accessToken: String
    var refreshToken: String
    var expiresAt: TimeInterval // unix seconds
    var user: AuthUser

    enum CodingKeys: String, CodingKey {
        case accessToken = "access_token"
        case refreshToken = "refresh_token"
        case expiresAt = "expires_at"
        case expiresIn = "expires_in"
        case user
    }

    init(accessToken: String, refreshToken: String, expiresAt: TimeInterval, user: AuthUser) {
        self.accessToken = accessToken
        self.refreshToken = refreshToken
        self.expiresAt = expiresAt
        self.user = user
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        accessToken = try c.decode(String.self, forKey: .accessToken)
        refreshToken = try c.decode(String.self, forKey: .refreshToken)
        user = try c.decode(AuthUser.self, forKey: .user)
        if let at = try c.decodeIfPresent(TimeInterval.self, forKey: .expiresAt) {
            expiresAt = at
        } else {
            let expIn = (try c.decodeIfPresent(TimeInterval.self, forKey: .expiresIn)) ?? 3600
            expiresAt = Date().timeIntervalSince1970 + expIn
        }
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(accessToken, forKey: .accessToken)
        try c.encode(refreshToken, forKey: .refreshToken)
        try c.encode(expiresAt, forKey: .expiresAt)
        try c.encode(user, forKey: .user)
    }

    var isExpiringSoon: Bool {
        expiresAt - Date().timeIntervalSince1970 < 60
    }
}

struct AuthUser: Codable, Equatable {
    var id: String
    var email: String?
}

/// Строка таблицы profiles (читается под RLS своим токеном).
struct Profile: Codable, Equatable {
    var id: String
    var email: String?
    var role: String?            // "student" | "teacher"
    var firstName: String?
    var lastName: String?
    var studentGrade: Int?
    var teacherType: String?
    var profileCompleted: Bool?

    enum CodingKeys: String, CodingKey {
        case id, email, role
        case firstName = "first_name"
        case lastName = "last_name"
        case studentGrade = "student_grade"
        case teacherType = "teacher_type"
        case profileCompleted = "profile_completed"
    }

    var isTeacher: Bool { role == "teacher" }

    /// Профиль требует completion-шага (как редирект на google_complete.html):
    /// нет валидной роли либо явный profile_completed=false после OAuth-signup.
    var needsCompletion: Bool {
        let r = (role ?? "").trimmingCharacters(in: .whitespaces)
        if r != "student" && r != "teacher" { return true }
        return profileCompleted == false
    }

    var displayName: String {
        let name = [firstName, lastName].compactMap { $0 }.joined(separator: " ")
        if !name.trimmingCharacters(in: .whitespaces).isEmpty { return name }
        if let email, let local = email.split(separator: "@").first { return String(local) }
        return "Пользователь"
    }

    /// Имя для start_homework_attempt — как inferNameFromUser в вебе.
    var hwStudentName: String {
        if let f = firstName, !f.trimmingCharacters(in: .whitespaces).isEmpty { return f }
        if let email, let local = email.split(separator: "@").first { return String(local) }
        return "Ученик"
    }
}
